File: sharedUI/src/commonMain/kotlin/com/calypsan/listenup/client/features/shell/components/ConnectionHealthBanner.kt
package com.calypsan.listenup.client.features.shell.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AccountCircle
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Info
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.calypsan.listenup.client.presentation.connection.ConnectionHealthUi
import listenup.composeapp.generated.resources.Res
import listenup.composeapp.generated.resources.common_dismiss
import listenup.composeapp.generated.resources.shell_session_lapsed_body
import listenup.composeapp.generated.resources.shell_session_lapsed_sign_in
import listenup.composeapp.generated.resources.shell_session_lapsed_title
import listenup.composeapp.generated.resources.shell_update_available_body
import listenup.composeapp.generated.resources.shell_update_available_title
import org.jetbrains.compose.resources.stringResource

/**
 * Shell-level connection-health banner. Renders one pill per non-[ConnectionHealthUi.Hidden]
 * state: session-lapse ("Signed out — sign in to sync" + Sign-in) and contract-version-skew
 * ("Update available" + Dismiss). Deliberately has no unreachable-server state — offline-first
 * means there's no ambient "offline" banner; connectivity surfaces only at point of need (book
 * detail, player), not as shell chrome.
 *
 * Hard rules (spec §10): never modal, never blocks navigation/playback/browse; no
 * auto-navigation on state entry — the only navigation is the user tapping Sign in.
 * Dismissal is UI-local and resets when the state next changes (`remember(state)`); the
 * [ConnectionHealthUi.Outdated] case additionally offers a persisted dismissal via [onDismiss].
 * Visual language follows the book-detail OfflineBanner (color-container pill).
 */
@Composable
fun ConnectionHealthBanner(
    state: ConnectionHealthUi,
    onSignIn: () -> Unit,
    onDismiss: () -> Unit,
    modifier: Modifier = Modifier,
) {
    var dismissed by remember(state) { mutableStateOf(false) }
    if (dismissed) return

    when (state) {
        ConnectionHealthUi.Hidden -> {
            return
        }

        ConnectionHealthUi.SessionExpired -> {
            BannerPill(
                modifier = modifier,
                container = MaterialTheme.colorScheme.errorContainer,
                onContainer = MaterialTheme.colorScheme.onErrorContainer,
                icon = Icons.Default.AccountCircle,
                title = stringResource(Res.string.shell_session_lapsed_title),
                body = stringResource(Res.string.shell_session_lapsed_body),
                actionLabel = stringResource(Res.string.shell_session_lapsed_sign_in),
                onAction = onSignIn,
                onClose = { dismissed = true },
            )
        }

        is ConnectionHealthUi.Outdated -> {
            BannerPill(
                modifier = modifier,
                container = MaterialTheme.colorScheme.tertiaryContainer,
                onContainer = MaterialTheme.colorScheme.onTertiaryContainer,
                icon = Icons.Default.Info,
                title = stringResource(Res.string.shell_update_available_title),
                body =
                    stringResource(
                        Res.string.shell_update_available_body,
                        state.clientVersion,
                        state.serverVersion,
                    ),
                actionLabel = stringResource(Res.string.common_dismiss),
                onAction = onDismiss,
                onClose = null,
            )
        }
    }
}

/**
 * Shared pill scaffold for [ConnectionHealthBanner]'s visible states: a rounded [Surface] with a
 * leading icon, title/body text column, a primary action [Button], and an optional trailing
 * dismiss [IconButton] ([onClose] is `null` when the primary action already is the dismissal).
 */
@Composable
private fun BannerPill(
    container: Color,
    onContainer: Color,
    icon: ImageVector,
    title: String,
    body: String,
    actionLabel: String,
    onAction: () -> Unit,
    onClose: (() -> Unit)?,
    modifier: Modifier = Modifier,
) {
    Surface(
        modifier = modifier.fillMaxWidth(),
        shape = RoundedCornerShape(24.dp),
        color = container,
        shadowElevation = 4.dp,
    ) {
        Row(
            modifier = Modifier.padding(horizontal = 16.dp, vertical = 10.dp),
            horizontalArrangement = Arrangement.spacedBy(12.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Icon(
                imageVector = icon,
                contentDescription = null,
                tint = onContainer,
            )
            Column(modifier = Modifier.weight(1f)) {
                Text(
                    text = title,
                    style = MaterialTheme.typography.titleSmall,
                    fontWeight = FontWeight.SemiBold,
                    color = onContainer,
                )
                Text(
                    text = body,
                    style = MaterialTheme.typography.bodySmall,
                    color = onContainer.copy(alpha = 0.85f),
                )
            }
            Button(
                onClick = onAction,
                colors =
                    ButtonDefaults.buttonColors(
                        containerColor = onContainer,
                        contentColor = container,
                    ),
            ) {
                Text(actionLabel)
            }
            if (onClose != null) {
                IconButton(onClick = onClose) {
                    Icon(
                        imageVector = Icons.Default.Close,
                        contentDescription = stringResource(Res.string.common_dismiss),
                        tint = onContainer,
                    )
                }
            }
        }
    }
}

File: app/iosApp/ListenUp/DesignSystem/Components/ConnectionHealthBanner.swift
import SwiftUI

/// Non-blocking connection-health banner for the mounted shell: "Update available" (a meaningful
/// client/server version skew, with Dismiss). Never modal — browsing and playback stay fully usable
/// underneath (Never Stranded). A Liquid-Glass card matching `SessionLapsedBanner`.
///
/// `sessionExpired` is intentionally NOT rendered here — it has its own `SessionLapsedBanner` driven
/// by `AuthStateObserver`, shown in the app root's `.sessionLapsed` branch. There is deliberately no
/// ambient "offline" state either: ListenUp is offline-first, and connectivity is surfaced only at
/// point of need (book detail, player).
struct ConnectionHealthBanner: View {
    let kind: ConnectionHealthKind
    var onDismiss: () -> Void

    var body: some View {
        switch kind {
        case .outdated(let clientVersion, let serverVersion):
            card(
                icon: "arrow.down.circle",
                title: Text("shell.update_available_title"),
                message: Text(updateBody(clientVersion, serverVersion)),
                actionLabel: Text("common.dismiss"),
                onAction: onDismiss
            )
        case .hidden, .sessionExpired:
            EmptyView()
        }
    }

    private func updateBody(_ client: String, _ server: String) -> String {
        String(format: String(localized: "shell.update_available_body"), client, server)
    }

    @ViewBuilder
    private func card(
        icon: String,
        title: Text,
        message: Text,
        actionLabel: Text,
        onAction: @escaping () -> Void
    ) -> some View {
        HStack(spacing: 12) {
            Image(systemName: icon)
                .font(.title3)
                .foregroundStyle(Color.listenUpOrange)

            VStack(alignment: .leading, spacing: 2) {
                title
                    .font(.subheadline.weight(.semibold))
                    .foregroundStyle(.primary)
                message
                    .font(.caption)
                    .foregroundStyle(.secondary)
                    .fixedSize(horizontal: false, vertical: true)
            }

            Spacer(minLength: 8)

            Button(action: onAction) {
                actionLabel
                    .font(.subheadline.weight(.semibold))
                    .foregroundStyle(Color.listenUpOrange)
                    .frame(minHeight: 44)
            }
            .buttonStyle(.plain)
        }
        .padding(14)
        .frame(maxWidth: .infinity, alignment: .leading)
        .glassControl(in: RoundedRectangle(cornerRadius: 14, style: .continuous))
        .padding(.horizontal, 16)
    }
}
